File: Editorial/0129_Sum_Root_to_Leaf_Numbers/CPP/2_Recursive_Preorder_Traversal.cpp
class Solution {
public:
    void preorder(TreeNode* r, int currNumber, int& rootToLeaf) {
        if (r != NULL) {
            currNumber = currNumber * 10 + r->val;
            // if it's a leaf, update root-to-leaf sum
            if (r->left == NULL && r->right == NULL) rootToLeaf += currNumber;
            preorder(r->left, currNumber, rootToLeaf);
            preorder(r->right, currNumber, rootToLeaf);
        }
    }
    int sumNumbers(TreeNode* root) {
        int rootToLeaf = 0;
        preorder(root, 0, rootToLeaf);
        return rootToLeaf;
    }
};

File: Editorial/0101-0200/0129_Sum_Root_to_Leaf_Numbers/TypeScript/2_Recursive_Preorder_Traversal.ts
function sumNumbers(root: TreeNode | null): number {
    function preorder(r: TreeNode | null, currNumber: number) {
        if (r) {
            currNumber = currNumber * 10 + r.val;
            // if it's a leaf, update root-to-leaf sum
            if (!r.left && !r.right) rootToLeaf += currNumber;
            preorder(r.left, currNumber);
            preorder(r.right, currNumber);
        }
    }
    let rootToLeaf = 0;
    preorder(root, 0);
    return rootToLeaf;
}
